### ISSUE TITLE
Elaborate the explanation of `constexpr` and add how to use `consteval`

### ISSUE THREAD

**Issue by hsutter** [isocpp/CppCoreGuidelines#1309]
See also #1307

**Comment by mbozzi** [comment]
Consider the sample code from #1307: 

>     int fcount = 0;
>     constexpr int f(int v)
>     {
>         if (v < 0)
>             ++fcount;
>         return v + v;
>     }
    
I think a rule should be added that discourages constexpr functions like `f`, because its *usage contract depends on the location of the call*.  To see why, consider the set of acceptable arguments.  If a call is evaluated outside of a constant expression, we can say `f(v)` for any `v`.  But, inside a constant expression, we're forbidden from saying `f(v)` for `v < 0`.

As far as I know, the contract of *some constexpr function **g** in a constant expression* - call it **g'** - can never (?) be wider than that of **g**.  My opinion is that the contracts of **g** and **g'** should generally be *the same*, no matter where the calls appear; functions that have differing contracts in these circumstances are undesirable.

If we're forced to distinguish between **g** and **g'**, then we've sort of created a dual API, and the disadvantages of those are well-understood.  In the worst case, it is unclear which subset of argument values are acceptable in a constant expression, and it seems likely that such functions violate the single-responsibility principle.

I suggest the Guidelines recommend that a constexpr function should have the same contract at run-time as it does at compile-time.

I believe it's generally possible to refactor by extracting **g'**.  For example, the function `f` above can be refactored into a "compile-time part" and a "run-time part", and the result is clearly better:

    constexpr int f_prime(int v) 
    {
        return v + v;
    }
    
    int fcount;
    int f()
    {
        if (v < 0)
            ++fcount;
        return f_prime(v);
    } 

If such a rule was introduced, it might offer an opportunity to communicate that constexpr functions can indeed be called at runtime - which is a common misconception, in my experience.

Thoughts?

**Comment by cubbimew** [comment]
An easily overlooked issue with constexpr functions is that a useful constexpr function may be a function that is **never pure**. `std::swap` is `constexpr` in C++20 (see [[utility.swap]](http://eel.is/c++draft/utility.swap)), in order to make it possible to swap two local variables inside an outer constexpr function, which, in turn, may well be pure and running at compile time.

**Comment by mbozzi** [comment]
> Wouldn't that forbid functions with wide contracts to be constexpr (e.g. vector::at)? Is that intended?

No, it's not intended, but I understand where you're coming from - that comment needs work.

What I want to avoid is constexpr functions that don't make it obvious when they can be used in a constant expression.  After all, the documentation for `std::vector::at` will say "throws `out_of_range` when out-of-bounds", and it's understood what happens if control reaches `throw` inside a constant expression.

> Can you think of a more realistic example?

Consider the function `double my::sin(double)`.  

If without careful consideration, I qualified it `constexpr`, I'd obtain a function that might or might not work at all, and if it worked, it might not be clear which arguments are acceptable in a constant expression.

[p0553r1](https://wg21.link/p0533r1) proposes making (some of) &lt;cmath&gt; constexpr, and discusses the subtleties involved.  The important part is that careful design work is required, and that the "constexpr-acceptable" arguments need to be carefully managed so that compilation errors occur only where expected, and not (e.g.,) as a result of an implementation detail that touches a namespace-scope LUT in certain cases.

In constexpr functions, such implementation details affect the API - and need to be documented - or eliminated.

It is probably best if such details don't affect the interface: this is the crux of my suggestion.  This can be accomplished (e.g.) by moving them to another function, as I suggested above.  [p0595r1](http://wg21.link/p0595r1) suggests another approach that could resolve such issues, and maybe some more examples.
